lib/pure/os.nim

```diff
@@ -3217,6 +3184,48 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
+proc sameFileContent*(path1, path2: string; checkSize = false; bufferSize = 8192.Positive): bool {.
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns true if both pathname arguments refer to files with identical
+  ## binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var
+    a, b: File
+  if not open(a, path1): return false
+  if not open(b, path2):
+    close(a)
+    return false
+  if checkSize and getFileInfo(a).size != getFileInfo(b).size:
+    close(a)
+    close(b)
+    return false
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:
+    while true:
+      var readA = readBuffer(a, bufA, bufferSize)
+      var readB = readBuffer(b, bufB, bufferSize)
+      if readA != readB:
+        result = false
+        break
+      if readA == 0:
+        result = true
+        break
+      result = equalMem(bufA, bufB, readA)
+      if not result: break
+      if readA != bufferSize: break # end of file
```

Comment:
I'm not at all convinced that this is correct? what about EINTR? eg https://stackoverflow.com/questions/53245721/are-interrupt-signals-dispatched-during-fread-and-fwrite-library-calls; eg: https://www.geeksforgeeks.org/fread-function-in-c/

> Return Value: The function returns the number of objects read successfully. If an error occurs, the return value can be less than count.

IMO the only robust way is by checking whether you're at the EOF, via `feof` (or rather its nim cross platform wrapper)

* https://www.tutorialspoint.com/c_standard_library/c_function_feof.htm

Reply:
is this ok? https://github.com/nim-lang/Nim/pull/16008/commits/5b068cbc656730b7663f8595db9079a2229f0900

Reply:
> is this ok?

there are 2 things that are wrong:
* first, you should only do the check on a short read:
```nim
if readA != bufferSize:
  let enda = io.endOfFile(a)
  let endb = io.endOfFile(b)
  if enda != endb: result = false
  if enda or endb: break
```

* second, `io.endOfFile` is wrong, just filed https://github.com/nim-lang/Nim/issues/16011; maybe let's fix this first so your PR can use `io.endOfFile`, otherwise you'll introduce a regression until endOfFile is fixed?

from man fread:

>  If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

from man feof:

> The function feof() tests the end-of-file indicator for the stream pointed to by stream, returning non-zero if it is set.  The end-of-file indicator may be
     cleared by explicitly calling clearerr(), or as a side-effect of other operations, e.g. fseek().

Reply:
So the original code was kinda buggy already...
:(